### PR TITLE
Split FunctionParameterSyntax and related types

### PR DESCRIPTION
**Pull request by ahoppen**
FunctionParameterSyntax was used in a number of cases (function parameter, closure parameter and enum case parameter) and because it needs to satisfy all of them all its parameters are optional. Split it into three different types that have non-optional children.

rdar://106874808

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test macOS

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test macOS

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test macOS

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test Linux

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test Windows

File: Sources/SwiftParser/Parameters.swift [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+extension RawParameterClauseSyntax: RawParameterClauseTrait {}
+extension RawClosureParameterClauseSyntax: RawParameterClauseTrait {}
+extension RawEnumCaseParameterClauseSyntax: RawParameterClauseTrait {}
+
+// MARK: Parameter parsing
+
+extension Parser {
+  mutating func parseFunctionParameter() -> RawFunctionParameterSyntax {
```
Comment (bnbarham) [inline Sources/SwiftParser/Parameters.swift:57]: There's a fair bit of duplication between these. Did you find that made more sense then having options on a `parseParameterLikeDecl(...)` sort of function? `RawParameterTrait` could include an `init` that just has alllll the things, where we ignore params that don't make sense where necessary.

File: CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift [hunk 2]
```diff
@@ -727,14 +815,12 @@ public let DECL_NODES: [Node] = [
...
         name: "Colon",
-        kind: .token(choices: [.token(tokenKind: "ColonToken")]),
-        isOptional: true
+        kind: .token(choices: [.token(tokenKind: "ColonToken")])
       ),
       Child(
         name: "Type",
         kind: .node(kind: "Type"),
-        nameForDiagnostics: "type",
-        isOptional: true
```
Comment (bnbarham) [inline CodeGeneration/Sources/SyntaxSupport/DeclNodes.swift:737]: 🥳

File: Sources/SwiftParser/Parameters.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SwiftParser/Parameters.swift:57]: I wanted to get the compile-time safety of not forgetting to pass one of the required arguments and I think we can only reasonably to do was to duplicate it.

Comment (bnbarham) [reply Sources/SwiftParser/Parameters.swift:57]: You can still have separate `parseFunctionParameter` etc. for callers outside. It'd just be `fileprivate` within `Parameters`.

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test macOS

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test macOS

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/64697
https://github.com/apple/swift-format/pull/495
https://github.com/apple/swift-stress-tester/pull/229

@swift-ci Please test Windows

**Review by bnbarham — APPROVED** [review]

**Pull request merged**